### PR TITLE
[WIP] Add new Type System Version

### PR DESCRIPTION
**Pull request by marsishandsome**
### What problem does this PR solve? 
Current Type System is not compatible with JDBC and SparkJDBC, e.g.  for `TINYINT` JDBC will convert to `java.lang.Integer`, SparkJDBC will convert to `IntegerType`, but Tispark will convert to `LongType`.

### What is changed and how it works?
TiSpark Type System V1 will be compatible with SparkJDBC.

| TiDB Type            | JDBC                 | Spark JDBC         | TiSpark V0         | TiSpark V1         | diff |
| -------------------- | -------------------- | ------------------ | ------------------ | ------------------ | ---- |
| BIT(1)               | java.lang.Boolean    | BooleanType        | LongType           | BooleanType        | *    |
| BIT( > 1)            | byte[]               | BooleanType        | LongType           | BinaryType         | *    |
| TINYINT              | java.lang.Integer    | IntegerType        | LongType           | IntegerType        | *    |
| BOOLEAN              | java.lang.Boolean    | BooleanType        | LongType           | BooleanType        | *    |
| SMALLINT [UNSIGNED]  | java.lang.Integer    | IntegerType        | LongType           | IntegerType        | *    |
| MEDIUMINT [UNSIGNED] | java.lang.Integer    | IntegerType        | LongType           | IntegerType        | *    |
| INT                  | java.lang.Integer    | IntegerType        | LongType           | IntegerType        | *    |
| INT UNSIGNED         | java.lang.Long       | LongType           | LongType           | LongType           |      |
| BIGINT               | java.lang.Long       | LongType           | LongType           | LongType           |      |
| BIGINT UNSIGNED      | java.math.BigInteger | DecimalType(20, 0) | DecimalType(20, 0) | DecimalType(20, 0) |      |
| FLOAT                | java.lang.Float      | DoubleType         | DoubleType         | DoubleType         |      |
| DOUBLE               | java.lang.Double     | DoubleType         | DoubleType         | DoubleType         |      |
| DECIMAL              | java.math.BigDecimal | DecimalType        | DecimalType        | DecimalType        |      |
| DATE                 | java.sql.Date        | DateType           | DateType           | DateType           |      |
| DATETIME             | java.sql.Timestamp   | TimestampType      | TimestampType      | TimestampType      |      |
| TIMESTAMP            | java.sql.Timestamp   | TimestampType      | TimestampType      | TimestampType      |      |
| TIME                 | java.sql.Time        | TimestampType      | LongType           | TimestampType      | *    |
| YEAR                 | java.sql.Date        | DateType           | LongType           | DateType           | *    |
| CHAR                 | java.lang.String     | StringType         | StringType         | StringType         |      |
| VARCHAR              | java.lang.String     | StringType         | StringType         | StringType         |      |
| BINARY               | byte[]               | BinaryType         | BinaryType         | BinaryType         |      |
| VARBINARY            | byte[]               | BinaryType         | BinaryType         | BinaryType         |      |
| TINYBLOB             | byte[]               | BinaryType         | BinaryType         | BinaryType         |      |
| TINYTEXT             | java.lang.String     | StringType         | StringType         | StringType         |      |
| BLOB                 | byte[]               | BinaryType         | BinaryType         | BinaryType         |      |
| TEXT                 | java.lang.String     | StringType         | StringType         | StringType         |      |
| MEDIUMBLOB           | byte[]               | BinaryType         | BinaryType         | BinaryType         |      |
| MEDIUMTEXT           | java.lang.String     | StringType         | StringType         | StringType         |      |
| LONGBLOB             | byte[]               | BinaryType         | BinaryType         | BinaryType         |      |
| LONGTEXT             | java.lang.String     | StringType         | StringType         | StringType         |      |
| ENUM                 | java.lang.String     | StringType         | StringType         | StringType         |      |
| SET                  | java.lang.String     | StringType         | StringType         | StringType         |      |

### Check List 

Tests 

 - Unit test

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

**Comment by marsishandsome** [comment]
/run-all-tests MAVEN_OPTS=-Dcom.pingcap.tikv.type_system_version=0

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=1 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=0 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=1 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=1 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=1 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=1 mode=full

**Comment by marsishandsome** [comment]
/run-all-tests JVM_ARGS=-Dcom.pingcap.tikv.type_system_version=1 mode=full tikv=v3.0.2 tidb=v3.0.2 pd=v3.0.2

File: docs/type_system.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# TiSpark Type System
```
Comment (birdstorm) [inline docs/type_system.md]: Hey are we going to provide both TiSpark V0 and V1 for users?

File: tikv-client/src/main/java/com/pingcap/tikv/key/RowKey.java [hunk 2]
```diff
@@ -56,8 +56,10 @@ public static RowKey toRowKey(long tableId, TypedKey handle) {
     Object obj = handle.getValue();
     if (obj instanceof Long) {
       return new RowKey(tableId, (long) obj);
+    } else if (obj instanceof Integer) {
+      return new RowKey(tableId, ((Integer) obj).longValue());
     }
-    throw new TiExpressionException("Cannot encode row key with non-long type");
+    throw new TiExpressionException("Cannot encode row key with non-long or non-integer type");
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/key/RowKey.java]: how about `Cannot encode row key with non-integer type` directly?

File: tikv-client/src/main/java/com/pingcap/tikv/operation/transformer/Cast.java [hunk 3]
```diff
@@ -29,6 +35,43 @@ public Cast(DataType type) {
...
+  private void castV1(Object value, Row row, int pos) {
+    if (value == null) {
+      row.set(pos, targetDataType, null);
+      return;
+    }
+    if (targetDataType instanceof IntegerType) {
+      Object casted;
+
+      String returnType = ((IntegerType) targetDataType).getV1ReturnJavaType();
+      if ("Long".equals(returnType)) {
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/operation/transformer/Cast.java]: `switch(returnType) { case ... }` ?

File: tikv-client/src/main/java/com/pingcap/tikv/types/IntegerType.java [hunk 4]
```diff
@@ -86,6 +86,83 @@ protected Object doConvertToTiDBType(Object value)
...
+        break;
+      default:
+        throw new TypeException("Invalid IntegerType flag: " + flag);
+    }
+
+    if (isUnsignedLong()) {
+      return unsignedValueOf(ret);
+    } else if (isUnsignedInt()) {
+      return ret;
+    } else if (this.getType() == MySQLType.TypeTiny) {
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/types/IntegerType.java]: switch case?

**Pull request closed without merging**